sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/connectionstring/StorageServiceConnectionString.java

```diff
@@ -0,0 +1,203 @@
+
+            if (isValidPrimarySecondaryPair(blobEndpoint, blobSecondaryEndpoint)
+                    && isValidPrimarySecondaryPair(queueEndpoint, queueSecondaryEndpoint)
+                    && isValidPrimarySecondaryPair(tableEndpoint, tableSecondaryEndpoint)
+                    && isValidPrimarySecondaryPair(fileEndpoint, fileSecondaryEndpoint)) {
+                return new StorageConnectionString(
+                        StorageAuthenticationSettings.fromConnectionSettings(settings),
+                        StorageEndpoint.fromStorageSettings(settings,
+                                "blob",
+                                Constants.ConnectionStringConstants.BLOB_ENDPOINT_NAME,
+                                Constants.ConnectionStringConstants.BLOB_SECONDARY_ENDPOINT_NAME,
+                                matchesAutomaticEndpointsSpec,
+                                logger),
+                        StorageEndpoint.fromStorageSettings(settings,
+                                "queue",
+                                Constants.ConnectionStringConstants.QUEUE_ENDPOINT_NAME,
+                                Constants.ConnectionStringConstants.QUEUE_SECONDARY_ENDPOINT_NAME,
+                                matchesAutomaticEndpointsSpec,
+                                logger),
+                        StorageEndpoint.fromStorageSettings(settings,
+                                "table",
+                                Constants.ConnectionStringConstants.TABLE_ENDPOINT_NAME,
+                                Constants.ConnectionStringConstants.TABLE_SECONDARY_ENDPOINT_NAME,
+                                matchesAutomaticEndpointsSpec,
+                                logger),
+                        StorageEndpoint.fromStorageSettings(settings,
+                                "file",
+                                Constants.ConnectionStringConstants.FILE_ENDPOINT_NAME,
+                                Constants.ConnectionStringConstants.FILE_SECONDARY_ENDPOINT_NAME,
+                                matchesAutomaticEndpointsSpec,
+                                logger),
+                        settings.getSettingValue(Constants.ConnectionStringConstants.ACCOUNT_NAME_NAME));
+            }
+        }
+        return null;
+    }
+
+    private static ConnectionSettingsFilter automaticEndpointsMatchSpec() {
+        return ConnectionSettingsFilter.matchesExactly(
+                ConnectionSettingsFilter.matchesAll(
+                        ConnectionSettingsFilter.matchesOne(
+                                ConnectionSettingsFilter.matchesAll(requireAccountKey()),
+                                requireSas()),
+                        requireAccountName(),
+                        optionalEndpoints(),
+                        optionalEndpointProtocolAndSuffix())
+        );
+    }
+
+    private static ConnectionSettingsFilter explicitEndpointsMatchSpec() {
```

Comment:
What's the purpose the this match? Any example fit in this one but failed the automaticEndpointsMatchSpec?

Reply:
Yes, if there is no `AccountName` but `BlobEndpoint` then automatic match will fail and explicit will succeeded.